pkg/ossm/ior.go

```diff
@@ -0,0 +1,177 @@
+package ossm
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/maistra/maistra-test-tool/pkg/util"
+)
+
+type RouteMetadata struct {
+	Name         string `json:"name"`
+	CreationTime string `json:"creationTimestamp"`
+}
+
+type Routes struct {
+	Metadata RouteMetadata `json:"metadata"`
+}
+
+func cleanupMultipleIOR() {
+	util.Log.Info("Delete the namespaces and smmr")
+	util.Shell(`../scripts/smmr/clean_members_50.sh`)
+	//Remove the gateways
+	util.Shell(`oc patch smcp/%s -n %s --type json -p='[{"op": "remove", "path": "/spec/gateways/additionalEgress"}]'`, smcpName, meshNamespace)
+	util.Shell(`oc patch smcp/%s -n %s --type json -p='[{"op": "remove", "path": "/spec/gateways/additionalIngress"}]'`, smcpName, meshNamespace)
+	time.Sleep(time.Duration(40) * time.Second)
+}
+
+// TestIOR tests IOR error regarding routes recreated: https://issues.redhat.com/browse/OSSM-1974. IOR will be deprecated on 2.4 and willl be removed on 3.0
+func TestIOR(t *testing.T) {
+	defer cleanupMultipleIOR()
+	//For 2.4 we need to enable IOR, by default should be disable
+	defer util.RecoverPanic(t)
+	smcpVersion, _ := util.ShellMuteOutput(`oc get smcp/%s -n %s -o jsonpath='{.spec.version}'`, smcpName, meshNamespace)
+	util.Log.Info("SMCP version: ", smcpVersion)
+	iorEnabled, _ := util.ShellMuteOutput(`oc get smcp/%s -n %s -o jsonpath='{.status.appliedValues.istio.gateways.istio-ingressgateway.ior_enabled}'`, smcpName, meshNamespace)
+	util.Log.Info("IOR enabled: ", iorEnabled)
+
+	// Check if IOR is enabled or not by default. Note: for >= 2.4 IOR is disabled by default, so we need to enable it in another subtest. For < 2.4 IOR is enabled by default.
+	t.Run("check_default_ior_state", func(t *testing.T) {
+		defer util.RecoverPanic(t)
+		if strings.Contains(smcpVersion, "v2.4") {
+			util.Log.Info("IOR should be disabled by default")
+			if iorEnabled == "true" {
+				t.Errorf("IOR should be disabled by default")
+			}
+		} else {
+			if iorEnabled == "false" {
+				t.Errorf("IOR should be enabled by default")
+			}
+		}
+	})
+	// Check if we can enable ior for >= 2.4. IOR is disabled by default.
+	t.Run("enable_ior", func(t *testing.T) {
+		defer util.RecoverPanic(t)
+		if strings.Contains(smcpVersion, "v2.4") {
+			util.Log.Info("IOR should be disabled by default")
+			if iorEnabled == "true" {
+				t.Errorf("IOR should be disabled by default")
+			} else {
+				//Need to enable ior on the smcp
+				util.Log.Info("Enabling IOR")
+				util.Shell(`oc patch smcp/%s -n %s --type json -p='[{"op": "add", "path": "/status/appliedValues/istio/gateways/istio-ingressgateway/ior_enabled", "value": true}]'`, smcpName, meshNamespace)
```

Comment:
No. No-one except the operator should (and can) modify the `status`.

@fjglira you need to patch the appropriate field in the `spec` .

Reply:
Oh, my mistake. I see it. Will change this

Reply:
@yannuil, please can you tell me in 2.4 which is the appropriate spec field for ior_enable? Because for 2.3 is not editable because he is only inside appliedValues:
`  appliedValues:
    istio:
      kiali:
        dashboard:
          viewOnlyMode: false
        enabled: true
        ingress:
          enabled: true
        install: true
        resourceName: kiali
      tracing:
        enabled: true
        ingress:
          enabled: true
        jaeger:
          install: true
          resourceName: jaeger
          template: all-in-one
        provider: jaeger
      global:
        proxy_init:
          image: proxy-init-rhel7
          resources:
            limits:
              cpu: 2000m
              memory: 1024Mi
            requests:
              cpu: 10m
              memory: 10Mi
        multiCluster:
          clusterName: ''
          enabled: false
          globalDomainSuffix: global
          includeEnvoyFilter: true
          multiClusterOverrides:
            expansionEnabled: null
            multiClusterEnabled: null
        sds:
          token:
            aud: istio-ca
        defaultNodeSelector: {}
        remotePilotAddress: ''
        imagePullSecrets: []
        defaultResources:
          requests:
            cpu: 10m
            memory: 128Mi
        operatorManageWebhooks: false
        mountMtlsCerts: false
        meshID: ''
        centralIstiod: false
        istioNamespace: istio-system
        pilotCertProvider: istiod
        tracer:
          datadog:
            address: '$(HOST_IP):8126'
          lightstep:
            accessToken: ''
            address: ''
          stackdriver:
            debug: false
            maxNumberOfAnnotations: 200
            maxNumberOfAttributes: 200
            maxNumberOfMessageEvents: 200
          zipkin:
            address: ''
        mtls:
          auto: true
          enabled: false
        network: ''
        proxy:
          autoInject: disabled
          logLevel: warning
          includeInboundPorts: '*'
          resources:
            limits:
              cpu: 2000m
              memory: 1024Mi
            requests:
              cpu: 10m
              memory: 128Mi
          includeIPRanges: '*'
          excludeOutboundPorts: ''
          readinessFailureThreshold: 30
          tracer: zipkin
          dnsRefreshRate: 300s
          excludeInboundPorts: ''
          statusPort: 15021
          privileged: false
          readinessPeriodSeconds: 2
          enableCoreDump: false
          readinessInitialDelaySeconds: 1
          excludeIPRanges: ''
          image: >-
            registry.redhat.io/openshift-service-mesh/proxyv2-rhel8@sha256:15444e6b93a05d88c6fc6ce1d1e7744635141291b86a7c3abb57758849a5d19e
          componentLogLevel: 'misc:error'
          clusterDomain: cluster.local
        k8sIngress:
          enableHttps: false
          enabled: false
          gatewayName: ingressgateway
        prometheusNamespace: istio-system
        omitSidecarInjectorConfigMap: false
        meshNetworks: {}
        meshExpansion:
          enabled: false
          useILB: false
        sts:
          servicePort: 0
        imagePullPolicy: ''
        useMCP: false
        istiod:
          enableAnalysis: false
        hub: registry.redhat.io/openshift-service-mesh
        oneNamespace: false
        arch:
          amd64: 2
          ppc64le: 2
          s390x: 2
        caAddress: ''
        oauthproxy:
          hub: registry.redhat.io/openshift4
          image: >-
            quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:b9d5f941b1c94c58f9224817e208a0bb595ab09a6f08e3766fac53ab9a0cb843
          imagePullPolicy: IfNotPresent
          tag: v4.4
        configRootNamespace: istio-system
        defaultConfigVisibilitySettings: []
        configNamespace: istio-system
        jwtPolicy: first-party-jwt
        tag: 2.3.2
        tls:
          cipherSuites: []
          ecdhCurves: []
          maxProtocolVersion: ''
          minProtocolVersion: ''
        configValidation: true
        defaultTolerations: []
        manageNetworkPolicy: true
        defaultPodDisruptionBudget:
          enabled: false
        trustDomain: ''
        logAsJson: false
        enableTracing: true
        externalIstiod: false
        priorityClassName: ''
        logging:
          level: 'default:warn'
      grafana:
        enabled: true
        image: >-
          registry.redhat.io/openshift-service-mesh/grafana-rhel8@sha256:6c4b1ec87fdb6259a9845ac8530b7a6ea249acc2ba205ed79a6aa9a639d1a18f
        ingress:
          enabled: true
        service:
          annotations:
            service.alpha.openshift.io/serving-cert-secret-name: grafana-tls
      sidecarInjectorWebhook:
        enableNamespacesByDefault: false
        objectSelector:
          enabled: false
      rateLimiting:
        rls:
          image: >-
            registry.redhat.io/openshift-service-mesh/ratelimit-rhel8@sha256:12aaa9c8c84214fbffbd6a0e193bb84b9ad6bb64ddecbe06e1700cd184218214
      mixer:
        adapters:
          prometheus:
            enabled: true
        policy:
          enabled: false
        telemetry:
          enabled: false
      prometheus:
        enabled: true
        image: >-
          registry.redhat.io/openshift-service-mesh/prometheus-rhel8@sha256:39224efbb0d80d1f59c290046c981008926a0bce90be25542958045938b9dfcd
        ingress:
          enabled: true
        resources:
          requests:
            cpu: 10m
            memory: 128Mi
        service:
          annotations:
            service.alpha.openshift.io/serving-cert-secret-name: prometheus-tls
      clusterResources: true
      gateways:
        enabled: true
        istio-egressgateway:
          autoscaleEnabled: false
          enabled: true
          gatewayType: egress
          name: istio-egressgateway
          resources:
            requests:
              cpu: 10m
              memory: 128Mi
        istio-ingressgateway:
          autoscaleEnabled: false
          enabled: true
          gatewayType: ingress
          ior_enabled: true`

Reply:
because that specific sub-test case is only to enable ior because by default is going to be disabled in 2.4

Reply:
Changed